src/detect-ikev1-exch-type.c

```diff
@@ -0,0 +1,287 @@
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ *
+ * \author Frank Honza <frank.honza@dcso.de>
+ */
+
+#include "suricata-common.h"
+#include "conf.h"
+#include "detect.h"
+#include "detect-parse.h"
+#include "detect-engine.h"
+#include "detect-engine-content-inspection.h"
+#include "detect-ikev1-exch-type.h"
+#include "app-layer-parser.h"
+#include "util-byte.h"
+
+#include "rust-bindings.h"
+
+/**
+ *   [ikev1.exchtype]:[<|>|<=|>=]<type>;
+ */
+#define PARSE_REGEX "^\\s*(<=|>=|<|>)?\\s*([0-9]+)\\s*$"
+static DetectParseRegex parse_regex;
+
+enum DetectIkev1ExchTypeCompareMode {
+    PROCEDURE_EQ = 1, /* equal */
+    PROCEDURE_LT, /* less than */
+    PROCEDURE_LE, /* less than or equal */
+    PROCEDURE_GT, /* greater than */
+    PROCEDURE_GE, /* greater than or equal */
+};
+
+typedef struct {
+    uint32_t exch_type;
+    enum DetectIkev1ExchTypeCompareMode mode;
+} DetectIkev1ExchTypeData;
+
+static DetectIkev1ExchTypeData *DetectIkev1ExchTypeParse (const char *);
+static int DetectIkev1ExchTypeSetup (DetectEngineCtx *, Signature *s, const char *str);
+static void DetectIkev1ExchTypeFree(DetectEngineCtx *, void *);
+static int g_ikev1_exch_type_buffer_id = 0;
+
+static int DetectEngineInspectIkev1ExchTypeGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id);
+
+static int DetectIkev1ExchTypeMatch (DetectEngineThreadCtx *, Flow *,
+                                   uint8_t, void *, void *, const Signature *,
+                                   const SigMatchCtx *);
+
+/**
+ * \brief Registration function for ikev1.exchtype keyword.
+ */
+void DetectIkev1ExchTypeRegister (void)
+{
+    sigmatch_table[DETECT_AL_IKEV1_EXCH_TYPE].name = "ikev1.exchtype";
+    sigmatch_table[DETECT_AL_IKEV1_EXCH_TYPE].desc = "match IKEv1 exchange type";
+    sigmatch_table[DETECT_AL_IKEV1_EXCH_TYPE].url = "/rules/ikev1-keywords.html#ikev1-exchtype";
+    sigmatch_table[DETECT_AL_IKEV1_EXCH_TYPE].AppLayerTxMatch = DetectIkev1ExchTypeMatch;
+    sigmatch_table[DETECT_AL_IKEV1_EXCH_TYPE].Setup = DetectIkev1ExchTypeSetup;
+    sigmatch_table[DETECT_AL_IKEV1_EXCH_TYPE].Free = DetectIkev1ExchTypeFree;
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex);
+
+    DetectAppLayerInspectEngineRegister("ikev1.exchtype",
+            ALPROTO_IKEV1, SIG_FLAG_TOSERVER, 1,
+            DetectEngineInspectIkev1ExchTypeGeneric);
+
+    DetectAppLayerInspectEngineRegister("ikev1.exchtype",
+            ALPROTO_IKEV1, SIG_FLAG_TOCLIENT, 1,
+            DetectEngineInspectIkev1ExchTypeGeneric);
+
+    g_ikev1_exch_type_buffer_id = DetectBufferTypeGetByName("ikev1.exchtype");
+}
+
+static int DetectEngineInspectIkev1ExchTypeGeneric(ThreadVars *tv,
```

Comment:
Is this needed ? (it is not a tick buffer keyword)

Reply:
I've tried to remove this function but got a segfault in this line: ```sigmatch_table[smd->type].Match(det_ctx, p, s, smd->ctx)``` in detect-engine.c

Have tried to do it like you did in detect-icmpv6-mtu.c by specifying the Match-function directly, but the function signature there has no transaction (which i need to obtain the value from rust).

Do you know an example somewhere?